cmd/client.go

```diff
@@ -242,6 +252,127 @@ type Config struct {
 	UploadLimit       int64
 	DownloadLimit     int64
 	Transport         http.RoundTripper
+	// Force TLS
+	forceTLS bool
+}
+
+// getCredsChain returns an []credentials.Provider array for the config
+// and the STS configuration (if present)
+func (config *Config) getCredsChain() ([]credentials.Provider, *probe.Error) {
+	var credsChain []credentials.Provider
+	// if an STS endpoint is set, we will add that to the chain
+	if stsEndpoint := env.Get("MC_STS_ENDPOINT_"+config.Alias, ""); stsEndpoint != "" {
+		// set AWS_WEB_IDENTITY_TOKEN_FILE is MC_WEB_IDENTITY_TOKEN_FILE is set
+		if val := env.Get("MC_WEB_IDENTITY_TOKEN_FILE_"+config.Alias, ""); val != "" {
+			os.Setenv("AWS_WEB_IDENTITY_TOKEN_FILE", val)
+			if val := env.Get("MC_ROLE_ARN_"+config.Alias, ""); val != "" {
+				os.Setenv("AWS_ROLE_ARN", val)
+			}
+			if val := env.Get("MC_ROLE_SESSION_NAME_"+config.Alias, randString(32, rand.NewSource(time.Now().UnixNano()), "mc-session-name-")); val != "" {
+				os.Setenv("AWS_ROLE_SESSION_NAME", val)
+			}
+		}
+		stsEndpointURL, err := url.Parse(stsEndpoint)
+		if err != nil {
+			return nil, probe.NewError(fmt.Errorf("Error parsing sts endpoint: %w", err))
+		}
+		config.forceTLS = stsEndpointURL.Scheme == "https"
```

Comment:
Sorry to bring this up again, but `config.forceTLS` shouldn't be in here. It should be parameter that is fed to `config.getTransport()`. You shouldn't store parameters as persistent state.

Reply:
Setting this makes the following code result in different transports:
```
credsChain, _ := cfg.getCredsChain()           // set config.forceTLS to true
transport, _ := cfg.getTransport()               // TLS based transport
```
opposed to:
```
transport, _ := cfg.getTransport()               // non-TLS based transport
credsChain, _ := cfg.getCredsChain()           // set config.forceTLS to true, but too late...
```
It's good to have separation of concerns, but in this case the order is important and that's non-trivial when looking at the code. The proper way to fix this is to let `getTransport()` check whether or not TLS is required.